File: test/extensions/dynamic_modules/test_data/go/network_integration_test/network_integration_test.go
package main

import (
	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterNetworkFilterConfigFactories(map[string]shared.NetworkFilterConfigFactory{
		"flow_control":     &flowControlConfigFactory{},
		"connection_state": &connectionStateConfigFactory{},
		"half_close":       &halfCloseConfigFactory{},
		"buffer_limits":    &bufferLimitsConfigFactory{},
	})
}

func main() {}

type flowControlConfigFactory struct {
	shared.EmptyNetworkFilterConfigFactory
}

func (f *flowControlConfigFactory) Create(shared.NetworkFilterConfigHandle, []byte) (shared.NetworkFilterFactory, error) {
	return &flowControlFactory{}, nil
}

type flowControlFactory struct {
	shared.EmptyNetworkFilterFactory
}

func (f *flowControlFactory) Create(handle shared.NetworkFilterHandle) shared.NetworkFilter {
	return &flowControlFilter{handle: handle}
}

type flowControlFilter struct {
	handle        shared.NetworkFilterHandle
	readsDisabled bool
	shared.EmptyNetworkFilter
}

func (f *flowControlFilter) OnRead(shared.NetworkBuffer, bool) shared.NetworkFilterStatus {
	if !f.readsDisabled {
		if status := f.handle.ReadDisable(true); status != shared.NetworkReadDisableStatusTransitionedToReadDisabled {
			panic("unexpected disable status")
		}
		if f.handle.ReadEnabled() {
			panic("expected reads to be disabled")
		}
		f.readsDisabled = true

		if status := f.handle.ReadDisable(false); status != shared.NetworkReadDisableStatusTransitionedToReadEnabled {
			panic("unexpected enable status")
		}
		if !f.handle.ReadEnabled() {
			panic("expected reads to be enabled")
		}
	}
	return shared.NetworkFilterStatusContinue
}

type connectionStateConfigFactory struct {
	shared.EmptyNetworkFilterConfigFactory
}

func (f *connectionStateConfigFactory) Create(shared.NetworkFilterConfigHandle, []byte) (shared.NetworkFilterFactory, error) {
	return &connectionStateFactory{}, nil
}

type connectionStateFactory struct {
	shared.EmptyNetworkFilterFactory
}

func (f *connectionStateFactory) Create(handle shared.NetworkFilterHandle) shared.NetworkFilter {
	return &connectionStateFilter{handle: handle}
}

type connectionStateFilter struct {
	handle shared.NetworkFilterHandle
	shared.EmptyNetworkFilter
}

func (f *connectionStateFilter) OnNewConnection() shared.NetworkFilterStatus {
	if f.handle.GetConnectionState() != shared.NetworkConnectionStateOpen {
		panic("expected open state on new connection")
	}
	return shared.NetworkFilterStatusContinue
}

func (f *connectionStateFilter) OnRead(shared.NetworkBuffer, bool) shared.NetworkFilterStatus {
	if f.handle.GetConnectionState() != shared.NetworkConnectionStateOpen {
		panic("expected open state on read")
	}
	return shared.NetworkFilterStatusContinue
}

func (f *connectionStateFilter) OnWrite(shared.NetworkBuffer, bool) shared.NetworkFilterStatus {
	if f.handle.GetConnectionState() != shared.NetworkConnectionStateOpen {
		panic("expected open state on write")
	}
	return shared.NetworkFilterStatusContinue
}

type halfCloseConfigFactory struct {
	shared.EmptyNetworkFilterConfigFactory
}

func (f *halfCloseConfigFactory) Create(shared.NetworkFilterConfigHandle, []byte) (shared.NetworkFilterFactory, error) {
	return &halfCloseFactory{}, nil
}

type halfCloseFactory struct {
	shared.EmptyNetworkFilterFactory
}

func (f *halfCloseFactory) Create(handle shared.NetworkFilterHandle) shared.NetworkFilter {
	return &halfCloseFilter{handle: handle}
}

type halfCloseFilter struct {
	handle shared.NetworkFilterHandle
	shared.EmptyNetworkFilter
}

func (f *halfCloseFilter) OnNewConnection() shared.NetworkFilterStatus {
	if !f.handle.IsHalfCloseEnabled() {
		panic("expected half-close to be enabled")
	}
	return shared.NetworkFilterStatusContinue
}

func (f *halfCloseFilter) OnRead(shared.NetworkBuffer, bool) shared.NetworkFilterStatus {
	f.handle.EnableHalfClose(false)
	if f.handle.IsHalfCloseEnabled() {
		panic("expected half-close to be disabled")
	}
	f.handle.EnableHalfClose(true)
	if !f.handle.IsHalfCloseEnabled() {
		panic("expected half-close to be re-enabled")
	}
	return shared.NetworkFilterStatusContinue
}

type bufferLimitsConfigFactory struct {
	shared.EmptyNetworkFilterConfigFactory
}

func (f *bufferLimitsConfigFactory) Create(shared.NetworkFilterConfigHandle, []byte) (shared.NetworkFilterFactory, error) {
	return &bufferLimitsFactory{}, nil
}

type bufferLimitsFactory struct {
	shared.EmptyNetworkFilterFactory
}

func (f *bufferLimitsFactory) Create(handle shared.NetworkFilterHandle) shared.NetworkFilter {
	return &bufferLimitsFilter{handle: handle}
}

type bufferLimitsFilter struct {
	handle shared.NetworkFilterHandle
	shared.EmptyNetworkFilter
}

func (f *bufferLimitsFilter) OnNewConnection() shared.NetworkFilterStatus {
	_ = f.handle.GetBufferLimit()
	f.handle.SetBufferLimits(32768)
	if f.handle.GetBufferLimit() != 32768 {
		panic("expected updated buffer limit")
	}
	return shared.NetworkFilterStatusContinue
}

File: test/extensions/dynamic_modules/test_data/cpp/network_integration_test.cc
#include <cassert>

#include "source/extensions/dynamic_modules/sdk/cpp/sdk_network.h"

namespace Envoy {
namespace DynamicModules {

class FlowControlFilter : public NetworkFilter {
public:
  explicit FlowControlFilter(NetworkFilterHandle& handle) : handle_(handle) {}

  NetworkFilterStatus onNewConnection() override { return NetworkFilterStatus::Continue; }

  NetworkFilterStatus onRead(NetworkBuffer&, bool) override {
    if (!reads_disabled_) {
      const auto disable_status = handle_.readDisable(true);
      assert(disable_status == NetworkReadDisableStatus::TransitionedToReadDisabled);
      assert(!handle_.readEnabled());
      reads_disabled_ = true;

      const auto enable_status = handle_.readDisable(false);
      assert(enable_status == NetworkReadDisableStatus::TransitionedToReadEnabled);
      assert(handle_.readEnabled());
    }
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onWrite(NetworkBuffer&, bool) override {
    return NetworkFilterStatus::Continue;
  }

  void onEvent(NetworkConnectionEvent) override {}
  void onDestroy() override {}

private:
  NetworkFilterHandle& handle_;
  bool reads_disabled_{false};
};

class FlowControlFactory : public NetworkFilterFactory {
public:
  std::unique_ptr<NetworkFilter> create(NetworkFilterHandle& handle) override {
    return std::make_unique<FlowControlFilter>(handle);
  }
};

class FlowControlConfigFactory : public NetworkFilterConfigFactory {
public:
  std::unique_ptr<NetworkFilterFactory> create(NetworkFilterConfigHandle&,
                                               std::string_view) override {
    return std::make_unique<FlowControlFactory>();
  }
};

REGISTER_NETWORK_FILTER_CONFIG_FACTORY(FlowControlConfigFactory, "flow_control");

class ConnectionStateFilter : public NetworkFilter {
public:
  explicit ConnectionStateFilter(NetworkFilterHandle& handle) : handle_(handle) {}

  NetworkFilterStatus onNewConnection() override {
    assert(handle_.getConnectionState() == NetworkConnectionState::Open);
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onRead(NetworkBuffer&, bool) override {
    assert(handle_.getConnectionState() == NetworkConnectionState::Open);
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onWrite(NetworkBuffer&, bool) override {
    assert(handle_.getConnectionState() == NetworkConnectionState::Open);
    return NetworkFilterStatus::Continue;
  }

  void onEvent(NetworkConnectionEvent) override {}
  void onDestroy() override {}

private:
  NetworkFilterHandle& handle_;
};

class ConnectionStateFactory : public NetworkFilterFactory {
public:
  std::unique_ptr<NetworkFilter> create(NetworkFilterHandle& handle) override {
    return std::make_unique<ConnectionStateFilter>(handle);
  }
};

class ConnectionStateConfigFactory : public NetworkFilterConfigFactory {
public:
  std::unique_ptr<NetworkFilterFactory> create(NetworkFilterConfigHandle&,
                                               std::string_view) override {
    return std::make_unique<ConnectionStateFactory>();
  }
};

REGISTER_NETWORK_FILTER_CONFIG_FACTORY(ConnectionStateConfigFactory, "connection_state");

class HalfCloseFilter : public NetworkFilter {
public:
  explicit HalfCloseFilter(NetworkFilterHandle& handle) : handle_(handle) {}

  NetworkFilterStatus onNewConnection() override {
    assert(handle_.isHalfCloseEnabled());
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onRead(NetworkBuffer&, bool) override {
    handle_.enableHalfClose(false);
    assert(!handle_.isHalfCloseEnabled());
    handle_.enableHalfClose(true);
    assert(handle_.isHalfCloseEnabled());
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onWrite(NetworkBuffer&, bool) override {
    return NetworkFilterStatus::Continue;
  }

  void onEvent(NetworkConnectionEvent) override {}
  void onDestroy() override {}

private:
  NetworkFilterHandle& handle_;
};

class HalfCloseFactory : public NetworkFilterFactory {
public:
  std::unique_ptr<NetworkFilter> create(NetworkFilterHandle& handle) override {
    return std::make_unique<HalfCloseFilter>(handle);
  }
};

class HalfCloseConfigFactory : public NetworkFilterConfigFactory {
public:
  std::unique_ptr<NetworkFilterFactory> create(NetworkFilterConfigHandle&,
                                               std::string_view) override {
    return std::make_unique<HalfCloseFactory>();
  }
};

REGISTER_NETWORK_FILTER_CONFIG_FACTORY(HalfCloseConfigFactory, "half_close");

class BufferLimitsFilter : public NetworkFilter {
public:
  explicit BufferLimitsFilter(NetworkFilterHandle& handle) : handle_(handle) {}

  NetworkFilterStatus onNewConnection() override {
    initial_buffer_limit_ = handle_.getBufferLimit();
    handle_.setBufferLimits(32768);
    assert(handle_.getBufferLimit() == 32768);
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onRead(NetworkBuffer&, bool) override {
    return NetworkFilterStatus::Continue;
  }

  NetworkFilterStatus onWrite(NetworkBuffer&, bool) override {
    return NetworkFilterStatus::Continue;
  }

  void onEvent(NetworkConnectionEvent) override {}
  void onDestroy() override {}

  void onAboveWriteBufferHighWatermark() override { above_high_watermark_called_ = true; }
  void onBelowWriteBufferLowWatermark() override { below_low_watermark_called_ = true; }

private:
  NetworkFilterHandle& handle_;
  inline static bool above_high_watermark_called_{false};
  inline static bool below_low_watermark_called_{false};
  inline static uint32_t initial_buffer_limit_{0};
};

class BufferLimitsFactory : public NetworkFilterFactory {
public:
  std::unique_ptr<NetworkFilter> create(NetworkFilterHandle& handle) override {
    return std::make_unique<BufferLimitsFilter>(handle);
  }
};

class BufferLimitsConfigFactory : public NetworkFilterConfigFactory {
public:
  std::unique_ptr<NetworkFilterFactory> create(NetworkFilterConfigHandle&,
                                               std::string_view) override {
    return std::make_unique<BufferLimitsFactory>();
  }
};

REGISTER_NETWORK_FILTER_CONFIG_FACTORY(BufferLimitsConfigFactory, "buffer_limits");

} // namespace DynamicModules
} // namespace Envoy
